x/crosschain/keeper/evm_hooks_test.go

```diff
@@ -2,95 +2,769 @@ package keeper_test
+		cctxList := k.GetAllCrossChainTx(ctx)
+		require.Len(t, cctxList, 0)
+	})
+	t.Run("error returned if unable to process an event", func(t *testing.T) {
+		k, ctx, sdkk, zk := keepertest.CrosschainKeeper(t)
+		k.GetAuthKeeper().GetModuleAccount(ctx, fungibletypes.ModuleName)
+
+		chain := common.BtcMainnetChain()
+		chainID := chain.ChainId
+		setSupportedChain(ctx, zk, chainID)
+		SetupStateForProcessLogs(t, ctx, k, zk, sdkk, chain)
+
+		block := GetValidZRC20WithdrawToBTC(t)
+		gasZRC20 := setupGasCoin(t, ctx, zk.FungibleKeeper, sdkk.EvmKeeper, chainID, "bitcoin", "BTC")
+		for _, log := range block.Logs {
+			log.Address = gasZRC20
+		}
+		ctx = ctx.WithChainID("test-21-1")
+
+		err := k.ProcessLogs(ctx, block.Logs, sample.EthAddress(), "")
+		require.ErrorContains(t, err, "ProcessZRC20WithdrawalEvent: failed to convert chainID")
+		cctxList := k.GetAllCrossChainTx(ctx)
+		require.Len(t, cctxList, 0)
+	})
+}
+
+func SetupStateForProcessLogsZetaSent(t *testing.T, ctx sdk.Context, k *crosschainkeeper.Keeper, zk keepertest.ZetaKeepers, sdkk keepertest.SDKKeepers, chain common.Chain) {
+	admin := sample.AccAddress()
+	setAdminPolicies(ctx, zk, admin)
+
+	assetAddress := sample.EthAddress().String()
+	gasZRC20 := setupGasCoin(t, ctx, zk.FungibleKeeper, sdkk.EvmKeeper, chain.ChainId, "ethereum", "ETH")
+	zrc20Addr := deployZRC20(
+		t,
+		ctx,
+		zk.FungibleKeeper,
+		sdkk.EvmKeeper,
+		chain.ChainId,
+		"ethereum",
+		assetAddress,
+		"ETH",
+	)
+	fungibleMsgServer := fungiblekeeper.NewMsgServerImpl(*zk.FungibleKeeper)
+	_, err := fungibleMsgServer.UpdateZRC20WithdrawFee(
+		sdk.UnwrapSDKContext(ctx),
+		fungibletypes.NewMsgUpdateZRC20WithdrawFee(admin, gasZRC20.String(), sdk.NewUint(withdrawFee), sdkmath.Uint{}),
+	)
+	require.NoError(t, err)
+	k.SetGasPrice(ctx, crosschaintypes.GasPrice{
+		ChainId:     chain.ChainId,
+		MedianIndex: 0,
+		Prices:      []uint64{gasPrice},
+	})
+	setupZRC20Pool(
+		t,
+		ctx,
+		zk.FungibleKeeper,
+		sdkk.BankKeeper,
+		zrc20Addr,
+	)
+}
+func SetupStateForProcessLogs(t *testing.T, ctx sdk.Context, k *crosschainkeeper.Keeper, zk keepertest.ZetaKeepers, sdkk keepertest.SDKKeepers, chain common.Chain) {
+
+	deploySystemContracts(t, ctx, zk.FungibleKeeper, sdkk.EvmKeeper)
+	tss := sample.Tss()
+	zk.ObserverKeeper.SetTSS(ctx, tss)
+	k.SetGasPrice(ctx, crosschaintypes.GasPrice{
+		ChainId: chain.ChainId,
+		Prices:  []uint64{100},
+	})
+
+	zk.ObserverKeeper.SetChainNonces(ctx, observertypes.ChainNonces{
+		Index:   chain.ChainName.String(),
+		ChainId: chain.ChainId,
+		Nonce:   0,
+	})
+	zk.ObserverKeeper.SetPendingNonces(ctx, observertypes.PendingNonces{
+		NonceLow:  0,
+		NonceHigh: 0,
+		ChainId:   chain.ChainId,
+		Tss:       tss.TssPubkey,
 	})
 }
 
 // receiver is 1EYVvXLusCxtVuEwoYvWRyN5EZTXwPVvo3
-func GetSampleBlockLegacyToAddress(t *testing.T) (receipt ethtypes.Receipt) {
+func GetInvalidZRC20WithdrawToExternal(t *testing.T) (receipt ethtypes.Receipt) {
```

Comment:
What about moving the block logic to a `testdata` package or in the `sample` package?

Reply:
837a063a193fcb2066c49b150b332af05d285887